pkg/file/redundancy/level.go

```diff
@@ -0,0 +1,176 @@
+// Copyright 2023 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package redundancy
+
+import (
+	"errors"
+	"fmt"
+
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+// Level is the redundancy level
+// which carries information about how much redundancy should be added to data to remain retrievable with a 1-10^(-6) certainty
+// in different groups of expected chunk retrival error rates (level values)
+type Level uint8
+
+const (
+	// no redundancy will be added
+	NONE Level = iota
+	// expected 1% chunk retrieval error rate
+	MEDIUM
+	// expected 5% chunk retrieval error rate
+	STRONG
+	// expected 10% chunk retrieval error rate
+	INSANE
+	// expected 50% chunk retrieval error rate
+	PARANOID
+)
+
+// NewLevel returns a Level coresponding to the passed number parameter
+// throws an error if there is no level for the passed number
+func NewLevel(n uint8) (Level, error) {
+	if n > uint8(PARANOID) {
+		return 0, fmt.Errorf("redundancy: number %d does not have corresponding level", n)
+	}
+	return Level(n), nil
+}
+
+// GetParities returns number of parities based on appendix F table 5
+func (l Level) GetParities(shards int) int {
+	et, err := l.getErasureTable()
+	if err != nil {
+		return 0
+	}
+	return et.getParities(shards)
+}
+
+// GetMaxShards returns back the maximum number of effective data chunks
+func (l Level) GetMaxShards() int {
+	p := l.GetParities(swarm.Branches)
+	return swarm.Branches - p
+}
+
+// GetEncParities returns number of parities for encrypted chunks based on appendix F table 6
+func (l Level) GetEncParities(shards int) int {
+	et, err := l.getEncErasureTable()
+	if err != nil {
+		return 0
+	}
+	return et.getParities(shards)
+}
+
+func (l Level) getErasureTable() (erasureTable, error) {
+	switch l {
+	case MEDIUM:
+		return mediumEt, nil
+	case STRONG:
+		return strongEt, nil
+	case INSANE:
+		return insaneEt, nil
+	case PARANOID:
+		return paranoidEt, nil
+	default:
+		return erasureTable{}, errors.New("redundancy: level NONE does not have erasure table")
```

Comment:
Since `Level` is an opaque type, any `unit8` can be cast into the `Level`, thus the default case won't always translate into the `NONE`. The `NONE` case should be handled separately from the default (unknown) case. The same is below.